corehq/form_processor/models.py

```diff
@@ -552,19 +552,43 @@ def xml_md5(self):
         return self.get_attachment_meta('form.xml').content_md5()
 
     def archive(self, user_id=None):
+        # If this archive was initiated by a user, delete all other stubs for this action so that this action
+        # isn't overridden
+        from couchforms.models import UnfinishedArchiveStub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
         if self.is_archived:
             return
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        FormAccessorSQL.archive_form(self, user_id=user_id)
-        xform_archived.send(sender="form_processor", xform=self)
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True) as archive_stub:
+            FormAccessorSQL.archive_form(self, user_id)
+            archive_stub.archive_history_updated()
+            xform_archived.send(sender="form_processor", xform=self)
 
     def unarchive(self, user_id=None):
+        # If this unarchive was initiated by a user, delete all other stubs for this action so that this action
+        # isn't overridden
+        from couchforms.models import UnfinishedArchiveStub
+        UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
         if not self.is_archived:
             return
-
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        FormAccessorSQL.unarchive_form(self, user_id=user_id)
-        xform_unarchived.send(sender="form_processor", xform=self)
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False) as archive_stub:
+            FormAccessorSQL.unarchive_form(self, user_id)
+            archive_stub.archive_history_updated()
+            xform_unarchived.send(sender="form_processor", xform=self)
+
+    def publish_archive_action_to_kafka(self, user_id, archive):
+        # Don't update the history, just send to kafka
+        from couchforms.models import UnfinishedArchiveStub
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        from corehq.form_processor.change_publishers import publish_form_saved
+        # Delete the original stub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        with unfinished_archive(instance=self, user_id=user_id, archive=archive):
+            xform_archived.send(sender="form_processor", xform=self)
```

Comment:
should be `xform_unarchived` when `archived == False`

Reply:
https://github.com/dimagi/commcare-hq/pull/22470/commits/d393107933168eeeb3f25f3319cb8d25e25bdbad